### ISSUE TITLE
ECH test inaccurate on Firefox?

### ISSUE THREAD

**Issue by nofxy** [privacytests/privacytests.org#225]
I see that ECH is marked as failing for Firefox, but in my own testing, using these sites [0] [1], they say it's working.

ECH has been enabled since version 119 [2], released nearly a year ago [3]. Is it possible the ECH test is failing in a way that misrepresents the browser's actual privacy features? 

[0] https://www.cloudflare.com/ssl/encrypted-sni/

[1] https://tls-ech.dev/

[2] https://support.mozilla.org/en-US/kb/faq-encrypted-client-hello#w_how-do-i-enable-ech-in-firefox

[3] https://www.mozilla.org/en-US/firefox/119.0/releasenotes/

**Comment by valenting** [comment]
Firefox resolves HTTPS records necessary for ECH either through DoH (enabled by default in US, CA, RU, UA), or through native DNS APIs (res_query)
However, on OSX this API is causing crashes: https://bugzilla.mozilla.org/show_bug.cgi?id=1882856

I assume the tests are being run on OSX? @arthuredelstein

**Comment by Anutrix** [comment]
I am on Windows with Firefox 135 and noticed something weird. Sometimes it gets stuck on plaintest and other times encrypted.

When page had loaded first, it said encrypted.

Note that I didn't do any thing except wait for some time and rerun the test. First it's encrypted, then it becomes plaintext.
![Image](https://github.com/user-attachments/assets/dfcb301a-2b1e-4353-b473-66a2a1a0a78a)

Once console starts returning plaintext, if I refresh the page, it becomes plaintext.
![Image](https://github.com/user-attachments/assets/ac4055f8-8d6e-4bc7-87c8-56eadc19cd90)
Stays that way on further refresh button clicks.

Then I held Shift key after some time and pressed refresh button and voila, it's back to encrypted.
![Image](https://github.com/user-attachments/assets/4e0cb073-78b1-4506-b07f-68d169dc8f7f)
It stays that way on further refreshes.

In summary:
Whenever I make the 2nd request with cache, the `fl` field in response doesn't change. In this case, it almost always returns plaintext.  
Whenever I make the request without cache(using Shift+Mouse click on refresh button), the `fl` field in response changes. In this case, it always returns encrypted.

**Comment by valenting** [comment]
Thanks for bringing that up.
I can reproduce the same. If I do a force refresh, it uses http/2 with ECH, but another refresh has it using http/3 without ECH.
I'll have a look on Monday

**Comment by Anutrix** [comment]
https://www.cloudflare.com/en-gb/ssl/encrypted-sni/
https://tls-ech.dev/
https://defo.ie/ech-check.php

Consistently passes on my Firefox on 3 sites including a Cloudflare page and official DEfO project https://defo.ie/.
They even provide test server links running on 6 popular servers

**Comment by valenting** [comment]
I think this issue can be closed now?

**Comment by JeremyRand** [comment]
Per this link: https://support.mozilla.org/en-US/kb/faq-encrypted-client-hello#w_how-will-ech-impact-enterprises-that-use-transparent-proxies

> ECH encryption is automatically disabled when proxies or middleboxes which are trusted by the browser are detected, so they remain unaffected.

My understanding from talking to Arthur (correct me if I'm mistaken) is that PrivacyTests uses a TLS intercepting proxy during testing. I wonder if the proxy's presence (more precisely, its CA cert's presence) is inducing Firefox to turn off ECH during testing. If so, then Firefox should be marked as failing the test, since disabling a privacy feature due to the presence of a root CA that isn't in Mozilla's root program is highly undesirable behavior from a privacy standpoint.

Another possibility is that the proxy itself is breaking ECH, e.g. it may be issuing TLS connections with ECH turned off. If that's the case, then it would be a bug for PrivacyTests to fix.

**Comment by Anutrix** [comment]
As per my comment above, if 5 different server tech on https://defo.ie/ pass on Firefox and official Cloudflare page always passes for Firefox, I don't see why this is considered a Firefox issue.

Unless you think privacytests someone trumps the result of all that.
Last I checked, the whole test relies on trying to copy what Cloudflare does on https://www.cloudflare.com/en-in/ssl/encrypted-sni/ by calling https://crypto.cloudflare.com/cdn-cgi/trace.
It either does this poorly or a bug needs to reported to Cloudflare that one of their security tools are giving wrong results risking users everywhere.

**Comment by JeremyRand** [comment]
> As per my comment above, if 5 different server tech on https://defo.ie/ pass on Firefox and official Cloudflare page always passes for Firefox, I don't see why this is considered a Firefox issue.

@Anutrix Because PrivacyTests' goal is to be *correct*, not to blindly copy incorrect behavior of other tests. It is still unclear why PrivacyTests' results differ from Cloudflare's results (I suggested an explanation above); thus it is not clear whether this is a bug in PrivacyTests or in other tests.

**Comment by Anutrix** [comment]
> > As per my comment above, if 5 different server tech on https://defo.ie/ pass on Firefox and official Cloudflare page always passes for Firefox, I don't see why this is considered a Firefox issue.
> 
> [@Anutrix](https://github.com/Anutrix) Because PrivacyTests' goal is to be _correct_, not to blindly copy incorrect behavior of other tests. It is still unclear why PrivacyTests' results differ from Cloudflare's results (I suggested an explanation above); thus it is not clear whether this is a bug in PrivacyTests or in other tests.

Except PrivacyTests literally tried to copy behavior of Cloudflare. Not sure who asked to match the behavior of other tests.
My opinion is that PrivacyTests should test against multiple implementations. Hopefully, few self-hosted ones if possible.
A good place to start is https://github.com/tlswg/draft-ietf-tls-esni/wiki/Implementations which is the implementation list by [Transport Layer Security Working Group](https://github.com/tlswg) includes https://defo.ie/ and Cloudflare.
If Firefox fails that, it fails that.

Looks like Firefox's implementation of ECH depends on DoH being enabled.
Source: https://support.mozilla.org/en-US/kb/firefox-dns-over-https#w_encrypted-client-hello-ech
It doesn't answer why it works sometimes but not other times but it is something to keep in mind.

Note that DoH is on by default in Firefox for US, Canada, Russia and Ukraine. This might be reason for the mixed experience by users.
Source: https://support.mozilla.org/en-US/kb/dns-over-https-doh-faqs#w_what-is-your-rollout-schedule
Based on this, PrivacyTests ECH Test is correct in every other country as PrivacyTests goals include running a plain default installation of the browser. However, it is practically wrong for users in these 4 countries.

On an unrelated note, Brave Browser always passes PrivacyTests somehow despite:
https://github.com/brave/brave-browser/issues/36205
And same case on my Android Device's latest Brave(latest version as of this comment):

**Comment by alfredonodo** [comment]
Hi,
any news on this?
Thank you

**Comment by alfredonodo** [comment]
> Per this link: https://support.mozilla.org/en-US/kb/faq-encrypted-client-hello#w_how-will-ech-impact-enterprises-that-use-transparent-proxies
> 
> > ECH encryption is automatically disabled when proxies or middleboxes which are trusted by the browser are detected, so they remain unaffected.
> 
> My understanding from talking to Arthur (correct me if I'm mistaken) is that PrivacyTests uses a TLS intercepting proxy during testing. I wonder if the proxy's presence (more precisely, its CA cert's presence) is inducing Firefox to turn off ECH during testing. If so, then Firefox should be marked as failing the test, since disabling a privacy feature due to the presence of a root CA that isn't in Mozilla's root program is highly undesirable behavior from a privacy standpoint.
> 
> Another possibility is that the proxy itself is breaking ECH, e.g. it may be issuing TLS connections with ECH turned off. If that's the case, then it would be a bug for PrivacyTests to fix.

Hello,
I tested latest firefox and brave version and ECH is working. The problem with firefox could be related to a proxy as you pointed out or [enterprise policy](https://mozilla.github.io/policy-templates/#disableencryptedclienthello) or a [ bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1725938). Is there any update?

**Comment by larseggert** [comment]
Go to about:logging and take a networking profile. Email to to necko@mozilla.com please.
